rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,555 @@
+where
+    G::NodeId: Eq + Hash,
+{
+    // The input graph
+    graph: G,
+    // The user-supplied mapping to use for swapping tokens
+    input_mapping: HashMap<G::NodeId, G::NodeId>,
+    // Number of trials
+    trials: usize,
+    // Seed for random selection of a node for a trial
+    rng_seed: Option<u64>,
+    // Map of NodeId to NodeIndex
+    node_map: HashMap<G::NodeId, NodeIndex>,
+    // Map of NodeIndex to NodeId
+    rev_node_map: HashMap<NodeIndex, G::NodeId>,
+}
+
+impl<G> TokenSwapper<G>
+where
+    G: NodeCount
+        + EdgeCount
+        + IntoEdges
+        + Visitable
+        + NodeIndexable
+        + IntoNeighborsDirected
+        + IntoNodeIdentifiers
+        + Send
+        + Sync,
+    G::NodeId: Hash + Eq + Send + Sync,
+{
+    fn new(
+        graph: G,
+        mapping: HashMap<G::NodeId, G::NodeId>,
+        trials: Option<usize>,
+        seed: Option<u64>,
+    ) -> Self {
+        TokenSwapper {
+            graph,
+            input_mapping: mapping,
+            trials: trials.unwrap_or(4),
+            rng_seed: seed,
+            node_map: HashMap::with_capacity(graph.node_count()),
+            rev_node_map: HashMap::with_capacity(graph.node_count()),
+        }
+    }
+
+    fn map(&mut self) -> Vec<Swap> {
+        let num_nodes = self.graph.node_count();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // Same as digraph with no self edges
+        let mut sub_digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // A list of nodes that are remaining to be tried
+        let mut todo_nodes = Vec::with_capacity(self.input_mapping.len());
+
+        // Add nodes to the digraph/sub_digraph and build the node maps
+        for node in self.graph.node_identifiers() {
+            let index = digraph.add_node(());
+            sub_digraph.add_node(());
+            self.node_map.insert(node, index);
+            self.rev_node_map.insert(index, node);
+        }
+        // The input mapping in HashMap form using NodeIndex
+        let mut tokens: HashMap<NodeIndex, NodeIndex> = self
+            .input_mapping
+            .iter()
+            .map(|(k, v)| (self.node_map[k], self.node_map[v]))
+            .collect();
+
+        // todo_nodes are all the mapping entries where left != right
+        for (node, dest) in &tokens {
+            if node != dest {
+                todo_nodes.push(*node);
+            }
+        }
+        // Add initial edges to the digraph/sub_digraph
+        for node in self.graph.node_identifiers() {
+            self.add_token_edges(
+                self.node_map[&node],
+                &mut digraph,
+                &mut sub_digraph,
+                &mut tokens,
+            );
+        }
+        // Do the trials
+        let seed = self.rng_seed;
```

Comment:
This will cause the same seed to be used on each of the trials which will result in each result to be the same. The issue is unlike in the serial case when there was a shared rng instance that would cycle based on the fixed seed, in this case we're initializing a thread local rng with the same seed value in each thread which will result in the same behavior across all the trials. The pattern I normally use to fix this is to create an outer rng to generate `self.trials` list of seeds and then use each of those seeds to initialize a thread local rng. For example in this case it would be:

```diff
diff --git a/rustworkx-core/src/token_swapper.rs b/rustworkx-core/src/token_swapper.rs
index 8184f7f..4239a43 100644
--- a/rustworkx-core/src/token_swapper.rs
+++ b/rustworkx-core/src/token_swapper.rs
@@ -122,10 +122,17 @@ where
             );
         }
         // Do the trials
-        let seed = self.rng_seed;
-        (0..self.trials)
+        let outer_rng: Pcg64 = match self.rng_seed {
+            Some(rng_seed) => Pcg64::seed_from_u64(rng_seed),
+            None => Pcg64::from_entropy(),
+        };
+        let seed_vec: Vec<u64> = outer_rng
+            .sample_iter(&rand::distributions::Standard)
+            .take(self.trials)
+            .collect();
+        seed_vec
             .into_par_iter()
-            .map(|_| {
+            .map(|seed| {
                 self.trial_map(
                     digraph.clone(),
                     sub_digraph.clone(),
@@ -190,15 +197,12 @@ where
         mut sub_digraph: StableGraph<(), (), Directed>,
         mut tokens: HashMap<NodeIndex, NodeIndex>,
         mut todo_nodes: Vec<NodeIndex>,
-        seed: Option<u64>,
+        seed: u64,
     ) -> Vec<Swap> {
         // Create a random trial list of swaps to move tokens to optimal positions
         let mut steps = 0;
         let mut swap_edges: Vec<Swap> = vec![];
-        let mut rng_seed: Pcg64 = match seed {
-            Some(rng_seed) => Pcg64::seed_from_u64(rng_seed),
-            None => Pcg64::from_entropy(),
-        };
+        let mut rng_seed: Pcg64 = Pcg64::seed_from_u64(seed);
         while !todo_nodes.is_empty() && steps <= 4 * digraph.node_count().pow(2) {
             // Choose a random todo_node
             let between = Uniform::new(0, todo_nodes.len());
```

Reply:
Ah, of course. Will fix.

Reply:
Done in https://github.com/Qiskit/rustworkx/pull/765/commits/805be9cdba948bf2f22830f844db6c6fd332056f. Still doing 2.6 sec.